src/bio/bio_buffer.c

```diff
@@ -551,6 +551,19 @@ iterate_biov(struct bio_desc *biod,
 				struct bio_copy_args *arg = data;
 
 				D_ASSERT(i < arg->ca_sgl_cnt);
+				/* re-check to avoid compiler warning
+				 * This function is weird in that it takes a callback and a void *
+				 * but then modifies the void * based on which callback is provided,
+				 * the compiler knows about what is behind the pointer in
+				 * different cases and the writes here would be a buffer overflow
+				 * for different cb functions.  The compiler does not like this and
+				 * only allows the code to compile based on the presence of the
+				 * assert() - and D_ASSERT() may be disabled so add an additional
+				 * call to assert() itself.  Overall the logic of modifying data
+				 * should be in the cb itself as it's not really a void * depsite
+				 * what it's declared as.
```

Comment:
I don't quite follow this comment, if it's about modifying the "void *" argument, why the assert(i < arg->ca_sgl_cnt) would make compiler happy? Maybe compiler just like to ensure there is no overflow issue?

Reply:
The compiler knows the size of data (it's on the stack in the parent) from different calls, my assumption is the compiler is working out the smallest size that data can point to and these writes are beyond that.  This code however is safe in that the function pointer is also passed in and these writes only happen for a certain pointer value so in practice it's safe.  The compiler does have a point though, this code is unusual in modifying a void * before calling the cb_fn.

Perhaps the fix is to have the cb_fn take `i` as an arg as well?  I can try pushing a patch for that in a bit.